Introduce ngram subsystem

vladimir-koshelev opened the pull request:
`NGram` subsystem consists of three main components:

- `NGramFileBasedIndex`  collects statistics about ngram rate in a project.  
See `NGramFileBasedIndex#getNumberOfOccurrences`.
- `NGramElementProvider` is an extension point for supporting other languages.
- `NGramContributor` weighs completion results according to the statistics.

bibaev committed:
Add example how to add custom weights for each lookup element

vladimir-koshelev committed:
Introduce a NGram weigher for a completion

bibaev commented on src/com/intellij/stats/ngram/NGramContributor.kt:
```diff
@@ -0,0 +1,54 @@
+
+    init {
+        val originalPosition = parameters.position
+        nGram = NGram.getNGramForElement(originalPosition)
```
Is it possible to perform this operation in O(1)? Seems, tree traversal on each completion could be too expensive for large files

vladimir-koshelev replied:
AFAIK on each completion, Idea makes copy the file with `IntelliJRulezzz` identifier, so I don't think that an additional traverse of the whole file on each completion is a big deal.

bibaev commented on src/com/intellij/stats/ngram/lang/JavaElementProvider.kt:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    override fun shouldIndex(element: PsiElement): Boolean {
+        return identifierRegex.matches(element.text)
```
`StringUtil.isJavaIdentifier(element.text)` looks a bit more properly

vladimir-koshelev replied:
Thanks! I've failed finding this method :-(

bibaev commented on src/com/intellij/stats/ngram/NGramContributor.kt:
```diff
@@ -0,0 +1,54 @@
+}
+
+
+private class NaiveNGramWeigher(parameters: CompletionParameters) : NGramWeigher(parameters,"naive ngram weigher") {
```
Let's name it consistently with other weighers in `camelCase`. There are weighers named: `templates`, `sameWords`, etc.

bibaev commented on src/com/intellij/stats/ngram/NGramContributor.kt:
```diff
@@ -0,0 +1,54 @@
+            sum += rate / (coefficient + 0.0)
+            currentNGram = currentNGram.dropHead()
+        }
+        return sum // put your score evaluation here
```
The comment is not valid more

bibaev commented on src/com/intellij/stats/ngram/NGramFileBasedIndex.kt:
```diff
@@ -0,0 +1,91 @@
+private object NGramIndexer : DataIndexer<NGram, Int, FileContent> {
+    override fun map(fileContent: FileContent): MutableMap<NGram, Int> {
+        if (ModuleUtil.findModuleForFile(fileContent.file, fileContent.project) == null) {
+            return HashMap()
```
`emptyMap()` could be used. But at first, replace return type with `Map<NGram, Int>`

bibaev commented on src/com/intellij/stats/ngram/NGramContributor.kt:
```diff
@@ -0,0 +1,54 @@
+        result.runRemainingContributors(parameters, { completionResult ->
+            val wrapped = CompletionResult.wrap(completionResult.lookupElement,
+                    completionResult.prefixMatcher,
+                    completionResult.sorter.weigh(NaiveNGramWeigher(parameters)))
```
We have to create a weigher before the `runRemainingContributors` call instead of creating a new weigher for each lookup element. This is my error. Please, extract the constructor call.

vladimir-koshelev committed:
Optimize indexing performance and cleanup code

- use AST Nodes instead of PSI elements;
- do not copy a node text for shouldIndex check;
- use StringUtil.isJavaIdentifier for shouldIndex check in
JavaElementProvider;
- minor cleanups.
